Library/Homebrew/test/cask/cask_loader_spec.rb

```diff
@@ -82,6 +82,36 @@
           (old_tap.path/"tap_migrations.json").write tap_migrations.to_json
         end
 
+        context "to a cask in an other tap" do
+          let(:old_tap) { Tap.fetch("homebrew", "foo") }
+          let(:new_tap) { Tap.fetch("homebrew", "bar") }
+
+          let(:cask_file) { new_tap.cask_dir/"#{token}.rb" }
+
+          before do
+            new_tap.cask_dir.mkpath
+            FileUtils.touch cask_file
+          end
+
+          it "does not warn when loading the short token" do
+            expect do
+              described_class.for(token)
+            end.not_to output.to_stderr
```

Comment:
This test fails. https://github.com/Homebrew/brew/pull/17385/files#diff-aea1b1bc211f85168f88057b630dac95a4947b4abfa99c158ac592090fb3b23dR460 calls https://github.com/Homebrew/brew/pull/17385/files#diff-aea1b1bc211f85168f88057b630dac95a4947b4abfa99c158ac592090fb3b23dR551

Reply:
I don't think that should be blocker but it sounds like it'd be nice to have.

Reply:
> If which is the cask, the old or the new?

The failing test is installing the cask without any tap name (the test calls it "short token"). So having a correct migration for cask `x` from tap `old` to `new`, having both taps enabled and executing `brew install x` writes a warning that is unnecessary and fails this test.

There are other tests (just below this) ensuring that `brew install new/x` returns no warning and `brew install old/x` returns a warning.

Reply:
After thinking this through, I just had a simple idea for a fix: Remove the warning when installing without tap name.

The warning is _Warning: Formula pe/old-tap/example was renamed to pe/new-tap/example._
The `FromNameLoader` is only used when the user specifies a cask/formula without a tap name (specifying a tap name uses `FromTapLoader`). If a user only specifies the cask/formula without a tap, the intention is to install from whatever tap it's at. Not caring if this cask was migrated. Hence, we just suppress the warning.

With this change, all three tests are green. Installing with the new tap or without tap prints no warning. Installing with the old tap prints a warning.

Did I miss anything in my reasoning? It seems too simple.

Reply:
Interesting, I was initially worried that turning off that warning would silence the normal rename message that we actually do want to display but it seems to work the same on this branch and master as well.

```console
$ brew info annie
...
Warning: Formula annie was renamed to lux.
...
```

The other case would be tap migration and rename which seems to work fine too so I think this should be good.

```console
$ brew info schismtracker
...
Warning: Formula schismtracker was renamed to homebrew/cask/schism-tracker.
Warning: Cask schismtracker was renamed to schism-tracker.
...
```

Reply:
> The other case would be tap migration and rename which seems to work fine too so I think this should be good.

I'm afraid it is broken. The warning you see comes from 
https://github.com/Homebrew/brew/blob/c0dd639c7fd68914de4bcaec290bf0edb387fd43/Library/Homebrew/cask/cask_loader.rb#L455
and only works for casks migrated to core. But I have made my change here
https://github.com/Homebrew/brew/blob/c0dd639c7fd68914de4bcaec290bf0edb387fd43/Library/Homebrew/cask/cask_loader.rb#L460

With my change, this warning is not printed anymore.

```shell
$ brew info example
Warning: Cask pe/old-tap/example was renamed to pe/new-tap/example.
…
```

So, we need to find an other solution.

Reply:
You're right. I did a poor job of testing that.

> So, we need to find another solution. Would it be somehow possible to only print warnings depending on where we are called from? Because as I understand it, we want the warnings in `brew info shortname` but not in `brew install shortname`.

Ideally the warnings would work like this:
- `brew info shortname` and `brew install shortname` print a warning when the shortname has changed (`oldname` -> `new name`)
- `brew info shortname` and `brew install shortname` don't print a warning when the shortname hasn't changed like when there is a tap migration but the token doesn't change

One way to make that work would be to add a `warn_on_rename` flag to the `.tap_cask_token_type` method. If `warn` is set, you warn every time. If it's not set but `warn_on_rename` is set, it warns when the token has changed. `FromTapLoader.tap_new` would have to add the same `warn_on_rename` flag so that it could pass it to the `.tap_cask_token_type` method. At least, that's the best I can come up with right now.

Reply:
> Edit: Or maybe there is value in always printing the tap migration warning even if a rename hasn't taken place. It's certainly more transparent.

I tried some fixes but nothing seemed easy and readable. So, I'll just let it be.